packages/core/src/hydration/event_replay.ts

```diff
@@ -79,3 +118,78 @@ export function setJSActionAttribute(
     }
   }
 }
+
+function setEventReplayer(dispatcher: Dispatcher) {
+  dispatcher.setEventReplayer(queue => {
+    for (const event of queue) {
+      handleEvent(event);
+    }
+  });
+}
+
+/**
+ * TODO: Update documentation here.
+ *
+ * @param target Component, Directive or DOM Node.
+ */
+export function getLViewByElement(target: any): LView|null {
+  let lView = readLViewData(target);
+  if (lView) {
+    return lView;
+  } else {
+    // if the context is not found then we need to traverse upwards up the DOM
+    // to find the nearest element that has already been monkey patched with data
+    let parent = target as any;
+    while (parent = parent.parentNode) {
+      lView = readLViewData(parent);
+
+      // the edge of the app was also reached here through another means
+      // (maybe because the DOM was changed manually).
+      if (lView) {
+        attachLViewId(target, lView);
+        return lView;
+      }
+    }
+  }
+  return null;
+}
+
+function handleEvent(event: EventInfoWrapper) {
+  const eventName = event.getEventType();
+  const nativeElement = event.getAction()!.element;
```

Comment:
Should we handle a case when `event.getAction()` returns undefined?
```suggestion
  const nativeElement = event.getAction()?.element;
```

Reply:
tom tells me that this is always populated.